### DISCUSSION TITLE
Automate the Super User creation

### DISCUSSION THREAD

**Discussion by simo735** [netbox-community/netbox#7283] [General]
Hello,

I am trying to automate the netbox installation with bash shell for linux server. And i have a problem with superuser creation.
The lines to do this action are:

```
source /opt/netbox/venv/bin/activate
python3 /opt/netbox/netbox/manage.py createsuperuser
```

The manage.py script is interactive and ask you for those information:
```
username (leave blank to use root):
email address: 
Password: 
Password (again): 
```

To interact with a program I tried some tips like  or EOF or printf to give it the answers as arguments , for example: 

`printf "root\ncontact@local.com\npassword123\npassword123\n" | python3 /opt/netbox/netbox/manage.py createsuperuser
`

But i get this error message:

`Superuser creation skipped due to not running in a TTY. You can run manage.py createsuperuser in your project to create one manually.`

Is there any tips to create the superuser without using TTY Terminal ?
Thank you in advance.

**Comment by kkthxbye-code** [comment]
Have a look at how netbox-docker creates the superuser:

https://github.com/netbox-community/netbox-docker/blob/e021390568b21f09015d54ee6e2504b20991affa/docker/docker-entrypoint.sh#L60-L66

**Reply by jeremystretch** [reply]
Yeah, calling `User.objects.create_superuser()` seems to be the recommended approach per Dango convention.

**Reply by simo735** [reply]
Thank you @kkthxbye-code & @jeremystretch it is exactly what I need.

**Comment by vyas-n** [comment]
There's now a `--no-input` flag you can pass into the createsuperuser command: https://docs.djangoproject.com/en/4.2/ref/django-admin/#cmdoption-createsuperuser-noinput
```bash
source /opt/netbox/venv/bin/activate
DJANGO_SUPERUSER_PASSWORD=<password> python3 /opt/netbox/netbox/manage.py createsuperuser --no-input --username <username> --email <email> 
```

**Reply by solarchemist** [reply]
Very nice tip, thank you! Unfortunately, the `createsuperuser` command is not idempotent (unless I am mistaken). If the same command (with the same user credentials) is run a second time (like might happen as part of an Ansible role) the command returns `"That username is already taken"`. If anyone knows how to use this command in an idempotent fashion, I'm all ears. Until then the `User.objects.create_superuser()` approach above remains the (only?) option.

**Reply by vyas-n** [reply]
@solarchemist For what it's worth, the `createsuperuser` solution is still idempotent (the user doesn't get overridden when the command is run again), it just throws a CLI error (exit code 1) when run multiple times.

You're correct that folks are likely to prefer the python script over this solution if they need to execute the script multiple times, but for folks that just need to automate running the task once (as I did for a helm chart), it's still a valid option.

**Reply by ji-podhead** [reply]
> There's now a `--no-input` flag you can pass into the createsuperuser command: https://docs.djangoproject.com/en/4.2/ref/django-admin/#cmdoption-createsuperuser-noinput
> 
> ```shell
> source /opt/netbox/venv/bin/activate
> DJANGO_SUPERUSER_PASSWORD=<password> python3 /opt/netbox/netbox/manage.py createsuperuser --no-input --username <username> --email <email> 
> ```

i tried this in ansible and the problem was that the container automatically migrates after it gets started.
   -  you cannot edit the db during this process and ansible is not waiting for the command to complete, nor is it outputting the results. 

i also cant simply restart the container and migrate manually. 
so now i will try to find the process and wait untill its finished, or , or see if i can stop the migration in the first place.
